Sources/AWSSDKSwiftCore/AWSShapes/Payload.swift

```diff
@@ -16,44 +16,107 @@ import struct Foundation.Data
 import NIO
 import NIOFoundationCompat
 
-/// Object storing request/response payload
+/// Holds a request or response payload. A request payload can be in the form of either a ByteBuffer or a stream function that will supply ByteBuffers to the HTTP client.
+/// A response payload only comes in the form of a ByteBuffer
 public struct AWSPayload {
-
-    /// Construct a payload from a ByteBuffer
-    public static func byteBuffer(_ byteBuffer: ByteBuffer) -> Self {
-        return Self(byteBuffer: byteBuffer)
+    
+    /// Internal enum
+    enum Payload {
+        case byteBuffer(ByteBuffer)
+        case stream(size: Int?, stream: (EventLoop)->EventLoopFuture<ByteBuffer>)
     }
-
-    /// Construct a payload from a Data
+    
+    internal let payload: Payload
+    
+    /// construct a payload from a ByteBuffer
+    public static func byteBuffer(_ buffer: ByteBuffer) -> Self {
+        return AWSPayload(payload: .byteBuffer(buffer))
+    }
+    
+    /// construct a payload from a stream function. If you supply a size the stream function will be called repeated until you supply the number of bytes specified. If you
+    /// don't supply a size the stream function will be called repeatedly until you supply an empty `ByteBuffer`
+    public static func stream(size: Int? = nil, stream: @escaping (EventLoop)->EventLoopFuture<ByteBuffer>) -> Self {
+        return AWSPayload(payload: .stream(size: size, stream: stream))
+    }
+    
+    /// Construct a payload from `Data`
     public static func data(_ data: Data) -> Self {
```

Comment:
I think we should give the option to inject a `ByteBufferAllocator` here. Default value is create a new one.

Reply:
done